### PR TITLE
EZP-31088: Refactored Location gateway methods to use Doctrine\DBAL

### PR DESCRIPTION
**Pull request by alongosz**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31088](https://jira.ez.no/browse/EZP-31088) blocking [EZP-30921](https://jira.ez.no/browse/EZP-30921)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `master (8.0@dev)` for eZ Platform `v3.0`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

This PR refactors the Legacy (Content) Location Gateway implementation (`\eZ\Publish\Core\Persistence\Legacy\Content\Location\Gateway\DoctrineDatabase`) to rely on `\Doctrine\DBAL\Connection` instead of our custom Database Handler.

**TODO**:
- [x] Refactor Content Location gateway methods to use Doctrine\DBAL.
- [x] Drop DatabaseHandler from Location GW DoctrineDatabase constructor.
- [x] Introduce strict types to Content Location GW method signatures.
- [x] Align other services which use now strictly-typed Location GW.
- [x] Replace `ezcontentobject_tree` literals with `Location\Gateway::CONTENT_TREE_TABLE` constant.
- [x] Mark Content Location gateway classes as final and internal.
- [x] Set database platform in Location GW constructor.
- [x] Simplify Location GW `DoctrineDatabase::listTrashed` method.
- [x] [PHPDoc] Add missing `@throws` attribute to various methods.
- [x] Drop obsolete `MAX_LIMIT` constant from Location GW `DoctrineDatabase`
- [x] Fix incorrect return variable name in `Location\Handler::create` method
- [x] [Tests] Fix mocking error in `LocationHandlerTest::testCreateLocation`
- [x] Align `Location\Gateway\DoctrineDatabase::getAllLocationsQueryBuilder` with `createNodeQueryBuilder`, remove one.
- [x] Optimize `Content\Location\Gateway\DoctrineDatabase::moveSubtreeNodes` method (time and cognitive complexity reduced, query optimization is too complex for this scope).
- [x] [CS] Align Content Location gateway class files headers.
- [x] [CS] Align Code Style with eZ Platform Code Style
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Wait for Travis.
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by alongosz** [comment]
> Is there any reason why for example `updateNodeAssignment` is now fully typehinted but `getBasicNodeData` not? I thought that we are also cleaning mixed for all ids in 3.0?

Good catch, seems I had missed some strict types in the `ExceptionConversion` class. Fixed via e649099.
I've looked it over again now, but should you find more misses please let me know.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 1]
```diff
@@ -1,313 +1,274 @@
...
+                'a',
+                'ezcontentobject',
+                'c',
+                $expr->andX(
+                    $expr->orX(
+                        $expr->eq(
+                            'a.contentobject_id',
+                            'c.id'
+                        )
+                    ),
```
[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]: Fixed via 49c5b9d532.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 2]
```diff
@@ -329,311 +290,318 @@ public function moveSubtreeNodes(array $sourceNodeData, array $destinationNodeDa
...
+                    $query->createPositionalParameter(
+                        array_filter(explode('/', $pathString)),
+                        Connection::PARAM_INT_ARRAY
                     )
                 )
             );
-        $statement = $query->prepare();
-        $statement->execute();
-        $hiddenSubtrees = $statement->fetchAll(\PDO::FETCH_COLUMN);
+        $count = (int)$query->execute()->fetchColumn(0);
```
[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]: > `0` is default value for the `$column` parameter

Hmm, I don't consider this a mistake, but fixed via 49c5b9d532 to be consistent with the codebase.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 3]
```diff
@@ -329,311 +290,318 @@ public function moveSubtreeNodes(array $sourceNodeData, array $destinationNodeDa
...
+            );
         }
     }
 
-    private function isHiddenByParent($pathString, array $rows)
+    private function getHiddenNodeIds(array $rows): array
     {
-        $parentNodeIds = explode('/', trim($pathString, '/'));
+        return array_map(
+            function (array $row) {
```
[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]: Fixed via 1d1e61f714.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 4]
```diff
@@ -329,311 +290,318 @@ public function moveSubtreeNodes(array $sourceNodeData, array $destinationNodeDa
...
+                return (int)$row['node_id'];
+            },
+            array_filter(
+                $rows,
+                function (array $row) {
```
[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]: Fixed via 1d1e61f714.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 5]
```diff
@@ -329,311 +290,318 @@ public function moveSubtreeNodes(array $sourceNodeData, array $destinationNodeDa
...
+                    return !empty($row['is_hidden']);
+                }
+            )
+        );
+    }
+
+    private function isHiddenByParent($pathString, array $hiddenNodeIds): bool
```
[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]: Fixed via 1d1e61f714.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 6]
```diff
@@ -329,311 +290,318 @@ public function moveSubtreeNodes(array $sourceNodeData, array $destinationNodeDa
...
-     */
-    public function updateSubtreeModificationTime($pathString, $timestamp = null)
+    private function moveSingleSubtreeNode(
+        int $nodeId,
+        array $sourceNodeData,
+        array $destinationNodeData,
+        string $newPathString,
+        string $newPathIdentificationString,
+        int $newParentId,
+        array $hiddenNodeIds
```
[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]: Fixed via 1d1e61f714.

**Comment by micszo** [comment]
Sanities OK.

**Pull request merged**